lib/class-wp-rest-image-editor-controller.php

```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
-		if ( empty( $image_path ) ) {
-			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
-		}
+		if ( isset( $params['rotation'] ) ) { // TODO: use `rotate` instead?
+			// CW / CCW
+			$result = $image_editor->rotate( 0 - $params['rotation'] );
 
-		$image_editor = wp_get_image_editor( $image_path );
-		if ( ! $image_editor->load() ) {
-			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+			// TODO: return the "real" error message?
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_rotation_failed', __( 'Unable to rotate this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		if ( isset( $params['rotation'] ) ) {
-			$image_editor->rotate( 0 - $params['rotation'] );
+		if ( isset( $params['x'], $params['y'], $params['width'], $params['height'] ) ) {
+			$size = $image_editor->get_size();
+
+			// Crop
+			$crop_x = round( ( $size['width'] * floatval( $params['x'] ) ) / 100.0 );
+			$crop_y = round( ( $size['height'] * floatval( $params['y'] ) ) / 100.0 );
+			$width  = round( ( $size['width'] * floatval( $params['width'] ) ) / 100.0 );
+			$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
+
+			$result = $image_editor->crop( $crop_x, $crop_y, $width, $height );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_crop_failed', __( 'Unable to crop this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		$size = $image_editor->get_size();
+		// Calculate the file name.
+		$image_ext  = pathinfo( $image_file, PATHINFO_EXTENSION );
+		$image_name = wp_basename( $image_file, ".{$image_ext}" );
+
+		// Do not append multiple `-edited` to the file name.
+		// The user may be editing a previously edited image.
+		if ( preg_match( '/-edited(-\d+)?$/', $image_name ) ) {
+			// Remove any `-1`, `-2`, etc. `wp_unique_filename()` will add the proper number.
+			$image_name = preg_replace( '/-edited(-\d+)?$/', '-edited', $image_name );
+		} else {
+			// Append `-edited` before the extension.
+			$image_name .= '-edited';
+		}
 
-		// Finally apply the modifications.
-		$crop_x = round( ( $size['width'] * floatval( $params['x'] ) ) / 100.0 );
-		$crop_y = round( ( $size['height'] * floatval( $params['y'] ) ) / 100.0 );
-		$width  = round( ( $size['width'] * floatval( $params['width'] ) ) / 100.0 );
-		$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
-		$image_editor->crop( $crop_x, $crop_y, $width, $height );
+		$filename = "{$image_name}.{$image_ext}";
 
-		// TODO: Generate filename based on edits.
-		$target_file = 'edited-' . $meta['original_name'];
+		// Create the uploads sub-directory if needed.
+		$uploads = wp_upload_dir();
 
-		$filename = rtrim( dirname( $image_path ), '/' ) . '/' . $target_file;
```

Comment:
Previously we were overriding the last edited file, right?

Reply:
Yeah, kind of. Was pretty messy. Overwriting the last edited file if it is in the same uploads subdirectory, but not all sub-sizes of it. If the edited file was smaller, and not all sub-sizes were created, some of the old sub-sizes were left "orphaned".

Better workflow and UX would be to keep all edited files, perhaps let the user know they can delete unused files from the media library.